File: calculator/src/main/java/at/tuwien/progspr16/calculator/interpreter/Calculator.java
package at.tuwien.progspr16.calculator.interpreter;

import org.antlr.v4.runtime.misc.Nullable;
import org.antlr.v4.runtime.tree.ParseTree;

import java.io.InputStream;
import java.io.PrintStream;
import java.util.Scanner;

public class Calculator implements Runnable {

    private final Scanner scanner;
    private final PrintStream out;
    private final CalculatorInterpretVisitor interpreter;

    public Calculator(InputStream in, PrintStream out) {
        this.scanner = new Scanner(in);
        this.out = out;
        this.interpreter = new CalculatorInterpretVisitor();
    }

    @Override
    public void run() {
        while (!Thread.currentThread().isInterrupted()) {
            String expr = nextCalculatorExpression();
            if (expr == null) break;

            interpret(expr);
        }
    }

    @Nullable
    private String nextCalculatorExpression() {
        out.print("Next calculator command: ");

        if (!scanner.hasNextLine()) return null;

        return scanner.nextLine();
    }

    private void interpret(String expr) {
        ParseTree tree = ParseUtils.parseCalculatorExpression(expr);

        interpreter.visit(tree);
    }
}

File: calculator/src/main/scala/at/tuwien/progspr16/calculator/interpreter/Calculator.scala
package at.tuwien.progspr16.calculator.interpreter

import java.io.{InputStream, PrintStream}
import java.util.Scanner

import at.tuwien.progspr16.calculator.grammar.CalculatorParser.BlockContext
import org.antlr.v4.runtime.ParserRuleContext
import org.antlr.v4.runtime.tree.ParseTree

class Calculator(in: InputStream = System.in, out: PrintStream = System.out) {
  val scanner: Scanner = new Scanner(in)

  private var active: Boolean = true
  private val interpreter: CalculatorInterpretVisitor = new CalculatorInterpretVisitor(this)
  def run(): Unit = {
    while (active) {
      val optNextLine = nextCalculatorExpression

      if (optNextLine.nonEmpty) {
        interpret(optNextLine.get)
      } else {
        setInactive()
      }
    }
  }

  private def nextCalculatorExpression: Option[String] = {
    out.print("--> ")
    if (!scanner.hasNextLine) Option.empty
    else Option(scanner.nextLine)
  }

  def setInactive(): Unit = active = false

  def interpret(expr: String): String = {
    val tree: ParseTree = ParseUtils.parseCalculatorExpression(expr)
    interpreter.visit(tree)
    interpreter.getCurrentState
  }

  def write(text: String) = out.print(text+ " ")

  def readInt() = scanner.nextInt()

  def readBlockCtx(): BlockContext = {
    val parseTree = ParseUtils.parseCalculatorExpression(scanner.next)
    parseTree.asInstanceOf[ParserRuleContext].children.get(0).asInstanceOf[BlockContext]
  }
}